Zend/zend_types.h

```diff
@@ -203,19 +203,30 @@ struct _zval_struct {
 	} u2;
 };
 
-typedef struct _zend_refcounted_h {
+typedef struct _zend_refcounted_common {
 	uint32_t         refcount;			/* reference counter 32-bit */
 	union {
 		struct {
 			ZEND_ENDIAN_LOHI_3(
 				zend_uchar    type,
-				zend_uchar    flags,    /* used for strings & objects */
-				uint16_t      gc_info)  /* keeps GC root number (or 0) and color */
+				zend_uchar    gc_flags,    /* GC flags, including GC color */
+				uint16_t      extra_flags) /* additional type specific flags */
```

Comment:
I don't like the Idea of mixing unchangeable flags with GC color. I can't say how this should look like, but probably this might be done in some better way.

On one hand, the GC algorithm would benefit if we change color to BLACK on ADDREF. This is not done now, but potentially might be done quite efficient (especially on 64-bit system) if we combine reference-counter and color mutation.

On the other hand, we may move GC color into GC buffer at all.

Reply:
ZEND_RC_DEBUG compatibility should be fixed.
ZEND_RC_MOD_CHECK() macro should use "u.v.gc_flags" instead of "u.v.flags".
Even better, would switch to macros.